crypto/dh/dh_key.c

```diff
@@ -95,17 +96,35 @@ static int compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
 
 int DH_compute_key(unsigned char *key, const BIGNUM *pub_key, DH *dh)
 {
+    int ret = 0, pad = 0;
+
+    /* compute the key; ret is constant unless compute_key is external */
 #ifdef FIPS_MODULE
-    return compute_key(key, pub_key, dh);
+    ret = compute_key(key, pub_key, dh);
 #else
-    return dh->meth->compute_key(key, pub_key, dh);
+    ret = dh->meth->compute_key(key, pub_key, dh);
 #endif
+    if (ret <= 0)
+        return ret;
+
+    /* count leading zero bytes */
```

Comment:
Why not do something like this (although it is not constant time either, it is at least not worse than the BN back and forth conversion):
```
    volatile size_t npad = 0;
    volatile size_t mask = 1;

    for (i = 0; i < ret; i++) {
        mask = mask & !buf[i];
        npad += mask;
    }
    memmove(buf, buf + npad, ret - npad);
    memset(buf + ret - npad, 0, npad);
    ret -= npad;
```

Reply:
Incorporated in latest force push. Thanks @t8m !